Memmap

alejoe91 opened the pull request:
This is to enable the creation of memmap arrays for sorting extractors for waveforms, pca scores, amplitudes, etc

- added self._tmp_folder (+set-get)
- modified set get features so that it does force conversion to in-memory numpy array

alejoe91 commented on spikeextractors/sortingextractor.py:
```diff
@@ -102,7 +112,7 @@ def set_unit_spike_features(self, unit_id, feature_name, value):
                     self._unit_features[unit_id] = {}
                 if isinstance(feature_name, str) and len(value) == len(self.get_unit_spike_train(unit_id)):
-                    self._unit_features[unit_id][feature_name] = np.asarray(value)
+                    self._unit_features[unit_id][feature_name] = value  # np.asarray(value) this is to keep memmap object
```
@colehurwitz any specific reason why this was cast as array? is it for lists?
I removed the cast to enable memmap arrays, but I can add a check for lists

colehurwitz replied:
I don't think there was much reason for casting it. Since spiketrains are numpy arrays, I think I wanted to keep the 1-1 correspondence between features and spike trains. I can't remember any other reason why.

alejoe91 replied:
ok then I'll keep this way. As long as they have the same length, it's ok

colehurwitz commented:
So the memmaping won't happen in the SortingExtractor class, but it will accept memmaped objects basically? How does this work?

alejoe91 commented:
> So the memmaping won't happen in the SortingExtractor class, but it will accept memmaped objects basically? How does this work?

This delas with dealing with the tmp_folder (deleted by the destructor). Temporary files can be saved in the `self._tmp_folder` by other functions. See this PR: https://github.com/SpikeInterface/spiketoolkit/pull/303

colehurwitz commented:
Hmm, do you think the logic for the memmap should be inside sortingextractor? So you can add features with "set_features" but specify memmap in that function?

alejoe91 commented:
there is no logic here, it's just a container in which we can save memmaped stuff. And it's deleted when the object is destructed (that's why it's convenient to have it here)

colehurwitz commented:
Ok, your solution is good!

colehurwitz commented:
Ready to merge?

samuelgarcia commented on spikeextractors/sortingextractor.py:
```diff
@@ -155,8 +165,13 @@ def get_unit_spike_features(self, unit_id, feature_name, start_frame=None, end_f
                             end_frame = np.inf
-                        spike_indices =  np.where(np.logical_and(spike_train >= start_frame, spike_train < end_frame))
-                        return self._unit_features[unit_id][feature_name][spike_indices]
+                        if start_frame == 0 and end_frame == np.inf:
```
Having inf for int is a bad idea. None sounds better. (or -1 but more dangerous)

alejoe91 replied:
in this case it's fine because it's only used internally with the np.where (so -1 won't work)

samuelgarcia replied:
Yes it is fine but we have to avoid this everywhere. No inf for indexing.

samuelgarcia commented:
I don't get the idea.
Where this self._tmp_folder will be used.
Why not having the same mechanism for recordingextractor and so the cacherecordingextractor would b ethe same ?

alejoe91 commented:
Ok I think this is ready @samuelgarcia

alejoe91 commented:
@samuelgarcia I added the creation of the folder and the allocate_array (also for the recordingextractor). I modified a bit the cache extractor to use the _tmp_folder.
Also in the allocate memory, one can initialize with an array (this will clean up spiketoolkit a lot!!)

samuelgarcia commented:
This is cool no ?

alejoe91 commented:
yes I think we can merger this
